cpp/ql/lib/semmle/code/cpp/models/implementations/Allocation.qll

```diff
@@ -373,6 +260,63 @@ private class NewArrayAllocationExpr extends AllocationExpr, NewArrayExpr {
   override predicate requiresDealloc() { not exists(this.getPlacementPointer()) }
 }
 
+/**
+ * Holds if `f` is an allocation function according to the
+ * extensible `allocationFunctionModel` predicate.
+ */
+private predicate isAllocationFunctionFromModel(
+  Function f, string namespace, string type, string name
+) {
+  exists(boolean subtypes | allocationFunctionModel(namespace, type, subtypes, name, _, _, _, _) |
+    if type = ""
+    then f.hasQualifiedName(namespace, "", name)
+    else
+      exists(Class c |
+        c.hasQualifiedName(namespace, type) and f.hasQualifiedName(namespace, _, name)
+      |
+        if subtypes = true
+        then f = c.getADerivedClass*().getAMemberFunction()
+        else f = c.getAMemberFunction()
+      )
+  )
+}
+
+/**
+ * An allocation function modeled via the extensible `allocationFunctionModel` predicate.
+ */
+private class AllocationFunctionFromModel extends AllocationFunction {
+  string namespace;
+  string type;
+  string name;
+
+  AllocationFunctionFromModel() { isAllocationFunctionFromModel(this, namespace, type, name) }
+
+  final override int getSizeArg() {
+    exists(string sizeArg |
+      allocationFunctionModel(namespace, type, _, name, sizeArg, _, _, _) and
```

Comment:
I'm trying to figure out if the `_` for `subtypes` is safe here or not.  Is it reasonable for there to be two models of the same function, one with subtypes and one without, and I guess differing in one of the last four parameters?  I think that would be an ambiguous model in the first place?

Reply:
Yes, that would be an ambiguous model. If we have something like:
```
["MyNamespace", "MyClass", False, "myAlloc", "0", "2", "", True]
["MyNamespace", "MyClass", False, "myAlloc", "1", "3", "", True]
```
we'd have one `AllocationFunction` whose `getSizeArg` predicate would return two values: `0` and `1`. That same `AllocationFunction` function would also have two results for `getMultArg`: `2` and `3`. This is not ideal, but since we have to stick to the interface offered by `AllocationFunction` we can't really do anything else